test/unit/org/apache/cassandra/index/sai/cql/AllowFilteringTest.java

```diff
@@ -0,0 +1,314 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.index.sai.cql;
+
+import org.junit.Test;
+
+import org.apache.cassandra.cql3.restrictions.StatementRestrictions;
+import org.apache.cassandra.index.sai.SAITester;
+import org.apache.cassandra.index.sai.StorageAttachedIndex;
+
+import static java.lang.String.format;
+import static org.junit.Assert.assertNotNull;
+
+/**
+ * Tests that {@code ALLOW FILTERING} is required only if needed.
+ */
+public class AllowFilteringTest extends SAITester
+{
+    @Test
+    public void testAllowFilteringOnFirstClusteringKeyColumn() throws Throwable
+    {
+        createTable("CREATE TABLE %s (k1 int, k2 int, c1 int, c2 int, c3 int, v1 int, " +
+                    "PRIMARY KEY ((k1, k2), c1, c2, c3))");
+        createIndex(format("CREATE CUSTOM INDEX ON %%s(c1) USING '%s'", StorageAttachedIndex.class.getName()));
```

Comment:
I have thought on ways of making it harder to forget the call to `waitForIndexQueryable`. However, this is tricky because:
* There is a `CQLTester#waitForIndex` method that we never use in SAI. Probably we should unify them and move all the logic for waiting on indexes to `CQLTester`.
* We have multiple tests where we create multiple indexes and then use a single common call to `waitForIndexQueryable`.
* We have multiple tests where we do other checks before doing the call to `waitForIndexQueryable`. That forces us to have both type of sync/async index creations, and we could fail distinguishing between them as easily as we forget calling `waitForIndexQueryable`. Maybe we could rename `createIndex` to `createIndexAsync` to remember us that the index could be incomplete or not queryable?
* The calls to `createIndex` sometimes embed a different table name in the index creation query. We would need to explicitly pass the table name to the equivalent `waitForIndexQueryable` call, and make sure that both the query and the table name arguments are consistent.

Since the improvement is not trivial and it's not related with the proposed changes, I would prefer to do it on a separate ticket.

There is also a bunch of known flakies on SAI that we have to fix before merging the feature branch, and there is a good chance that some of them are cause by forgetting to wait for the indexes. Maybe we can consider this improvement when fixing those flakies. wdyt? CC @mike-tr-adamson @maedhroz

Reply:
I think that autocompletion would show you both variants if we had them. Granted that is not bulletproof but it s better than nothing. A boolean to wait would also work. Or sthg along those lines yes.

Reply:
+1 on the `createIndexAsync` idea, and +1 to trying to unify our index wait mechanisms in a separate Jira

Reply:
I actually went through `SecondaryIndexManagerTest` and replaced `waitForIndex()` w/ `waitForIndexQueryable()`, and it seems to work just fine.

Reply:
I have created CASSANDRA-18521 for addressing those changes.